scripts/dwi_estimation_plot.py

```diff
@@ -0,0 +1,331 @@
+    bvecs = vertices
+    bvals = bval_shell * values
+
+    # Add a b0 value to the gradient table
+    bvals, bvecs = add_b0(bvals, bvecs)
+    return gradient_table(bvals, bvecs)
+
+
+def get_query_vectors(gtab, train_mask):
+    """Get the diffusion-encoding gradient vectors where the signal is to be estimated from the
+    gradient table and the training mask: the vectors of interest are those that are masked in
+    the training mask. b0 values are excluded."""
+
+    idx = np.logical_and(~train_mask, ~gtab.b0s_mask)
+    return gtab.bvecs[idx], np.where(idx)[0]
+
+
+def create_random_train_mask(gtab, size, seed=1234):
+    """Create a mask for the gradient table where a ``size`` number of indices will be
+    excluded. b0 values are excluded."""
+
+    rng = np.random.default_rng(seed)
+
+    # Get the indices of the non-zero diffusion-encoding gradient vector indices
+    nnzero_degv_idx = np.where(~gtab.b0s_mask)[0]
+
+    if nnzero_degv_idx.size < size:
+        raise ValueError(
+            f"Requested {size} values for masking; gradient table has {nnzero_degv_idx.size} "
+            "non-zero diffusion-encoding gradient vectors. Reduce the number of masked values."
+        )
+
+    lo = rng.choice(nnzero_degv_idx, size=size, replace=False)
+
+    # Exclude the b0s
+    zero_degv_idx = np.asarray(list(set(range(len(gtab.bvals))).difference(nnzero_degv_idx)))
+    lo = np.hstack([zero_degv_idx, lo])
+
+    train_mask = np.ones(len(gtab.bvals), dtype=bool)
+    train_mask[lo] = False
+
+    return train_mask
+
+
+def perform_experiment(gtab, S0, evals1, evecs, snr):
+    """Perform experiment: estimate the dMRI signal on a set of directions fitting a
+    Gaussian process to the rest of the data."""
+
+    # Fix the random number generator for reproducibility when generating the
+    # signal
+    seed = 1234
+    rng = np.random.default_rng(seed)
+
+    # Define the Gaussian process model parameters
+    kernel_model = "spherical"
+    lambda_s = 2.0
+    a = 1.0
+    sigma_sq = 0.5
+
+    # Define the Gaussian process model instance
+    gp_model = GaussianProcessModel(
+        kernel_model=kernel_model, lambda_s=lambda_s, a=a, sigma_sq=sigma_sq
+    )
+
+    # Create the DWI signal using a single tensor
+    signal = single_tensor(gtab, S0=S0, evals=evals1, evecs=evecs, snr=snr, rng=rng)
+
+    # Use all available data for training
+    gpfit = gp_model.fit(signal[~gtab.b0s_mask], gtab[~gtab.b0s_mask])
+
+    # Predict on an oversampled set of random directions over the unit sphere
+    # theta, phi = create_random_polar_coordinates(SAMPLING_DIRECTIONS, seed=seed)
+    # sph = Sphere(theta=theta, phi=phi)
+
+    # ToDo
+    # Not sure why all predictions are zero in gpfit.predict(sph.vertices)
+    # Also, when creating the convex hull, the gtab required is the one that
+    # would correspond to the new directions, so a new gtab would need to be
+    # generated
+    # return signal[~gtab.b0s_mask], gpfit.predict(sph.vertices)
```

Comment:
Not sure why this does not work (predictions are all 0's).

Reply:
Some investigation here:
- If I concatenate the desired 200 directions to the gtab directions used for the fitting and predict:
  ```
  _bvecs = np.vstack([gtab.bvecs[~gtab.b0s_mask], sph.vertices])
  gpfit.predict(_bvecs)
  ```
  it still results in all predictions corresponding to `sph.vertices` being 0.

  The issue arises from https://github.com/scikit-learn/scikit-learn/blob/f6871e80b1d11dfd18d3b85a9f7799c56371cb64/sklearn/gaussian_process/_gpr.py#L441. At that point the values in excess of the training mask are all zero. In the immediately previous line, the covariance matrix `K_trans = self.kernel_(self.X, self.X_train_)` the diagonal contains non-zero values (all valued 1e4: looks like the `sigma_sq_bounds` value).

  If I predict on `sph.vertices` (`gpfit.predict(sph.vertices)`), all entries in the covariance matrix are zero.

- If I keep 60 dirs ad the script input parameter to create the gradient table, and request less directions (`SAMPLING_DIRECTIONS = 20`, or `100`), `K_trans` is all zeros (largest value is 1e-62). 

 - Using 20, 30, 40 directions too create the gradient table, and predicting on `SAMPLING_DIRECTIONS = 100` does not result in zeros, but if I use 45 or 50 gradient table directions the prediction is all in zeros and `K_trans` is all zeros. Using 42 directions to create the gradient table, I already get values in the order of 1e-6 in the predictions.

   Same happens if I try to predict `SAMPLING_DIRECTIONS = 20` and use 42 directions for the gradient table: I get 1e-6 values in the prediction.

- Now if I set the parameter `a = 10` I get some non-zero predictions in `y_mean`. Have not checked the result in the plot.

So, this looks related to the kernel and the hyperparameter optimization ?

Additional not sure if the conditioning of the problem has anything to do with this.